rasa/core/policies/rule_policy.py

```diff
@@ -211,15 +241,132 @@ def _create_loop_unhappy_lookup_from_states(
+                # some action other than active_loop is predicted in unhappy path,
                 # therefore active_loop shouldn't be predicted by the rule
                 not is_prev_action_listen_in_state(states[-1])
-                and action not in {ACTION_LISTEN_NAME, active_loop}
+                and action != active_loop
             ):
                 lookup[feature_key] = DO_NOT_PREDICT_LOOP_ACTION
         return lookup
 
+    def _check_rule_restriction(self, rule_trackers):
+        rules_exceeding_max_user_turns = []
+        for tracker in rule_trackers:
+            number_of_user_uttered = sum(
+                isinstance(event, UserUttered) for event in tracker.events
+            )
+            if number_of_user_uttered > self.ALLOWED_NUMBER_OF_USER_INPUTS:
+                rules_exceeding_max_user_turns.append(tracker.sender_id)
+
+        if rules_exceeding_max_user_turns:
+            raise InvalidRule(
+                f"Found rules '{', '.join(rules_exceeding_max_user_turns)}' "
+                f"that contain more than {self.ALLOWED_NUMBER_OF_USER_INPUTS} "
+                f"user inputs. Rules are not meant to hardcode a state machine. "
+                f"Please use stories for these cases."
+            )
+
+    def _predict_next_action(
+        self,
+        tracker: TrackerWithCachedStates,
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter,
+    ) -> Optional[Text]:
+        probabilities = self.predict_action_probabilities(tracker, domain, interpreter)
+        # do not raise an error if RulePolicy didn't predict anything for stories;
+        # however for rules RulePolicy should always predict an action
+        if (
+            probabilities != self._default_predictions(domain)
+            or tracker.is_rule_tracker
+        ):
+            return domain.action_names[np.argmax(probabilities)]
+
+    def _check_prediction(
+        self,
+        tracker: TrackerWithCachedStates,
+        domain: Domain,
+        interpreter: NaturalLanguageInterpreter,
+        gold_action_name: Text,
+    ) -> Optional[Text]:
+
+        predicted_action_name = self._predict_next_action(tracker, domain, interpreter)
+        if not predicted_action_name or predicted_action_name == gold_action_name:
+            return
+
+        # RulePolicy will always predict active_loop first,
+        # but inside loop unhappy path there might be another action
+        if predicted_action_name == tracker.active_loop_name:
+            rasa.core.test.emulate_form_rejection(tracker)
+            predicted_action_name = self._predict_next_action(
+                tracker, domain, interpreter
+            )
+            if not predicted_action_name or predicted_action_name == gold_action_name:
+                return
+
+        tracker_type = "rule" if tracker.is_rule_tracker else "story"
+        return (
+            f"Action '{gold_action_name}' in {tracker_type} "
```

Comment:
Can we add which tracker is contradicting which other tracker?

Reply:
this is a problem, we don't know other tracker, we know that rules have learned smth different. If we want to know which rule predicted it, we need to significantly update tracker featurization

Reply:
or maybe not

Reply:
nope, it's not that easy, also rules could contain intersecting identical pieces